server/handler/steps.go

```diff
@@ -25,6 +25,34 @@ func GetStep(db *gorm.DB, w http.ResponseWriter, r *http.Request) {
 	respondJSON(w, http.StatusOK, step)
 }
 
+func CancelAllSteps(db *gorm.DB, engine *engine.Engine, w http.ResponseWriter, r *http.Request) {
+	steps := getAllSteps(db)
+	// first mark all non executing steps as cancelled
+	for _, aStep := range steps {
+		if len(aStep.Executions) == 0 {
+			db.Save(&model.Execution{
+				Status: model.Canceled,
+				StepID: aStep.ID,
+			})
+		}
+	}
+	// refresh steps
+	steps = getAllSteps(db)
+	// find currently running steps and cancel them
+	for _, aStep := range steps {
+		// check status of last one, there should not be any steps with no executions
+		if aStep.Executions[len(aStep.Executions)-1].Status == model.Started {
```

Comment:
There is an engine.getLatestExecution(db, step) function that you could use here..

Reply:
Sounds good, will make that function public and use it.